INTRODUCTION.md

```diff
@@ -461,29 +703,51 @@ A poll-based API is used to provide signaling back to the application,
 the application should call rd_kafka_poll() at regular intervals.
 The poll API will call the following configured callbacks (optional):
 
-  * message delivery report callback - signals that a message has been
-    delivered or failed delivery, allowing the application to take action
+  * `dr_msg_cb` - Message delivery report callback - signals that a message has
+    been delivered or failed delivery, allowing the application to take action
     and to release any application resources used in the message.
-  * error callback - signals an error. These errors are usually of an
-    informational nature, i.e., failure to connect to a broker, and the
+  * `error_cb` - Error callback - signals an error. These errors are usually of
+    an informational nature, i.e., failure to connect to a broker, and the
     application usually does not need to take any action.
     The type of error is passed as a rd_kafka_resp_err_t enum value,
     including both remote broker errors as well as local failures.
-
-
-Optional callbacks not triggered by poll, these may be called from any thread:
-
-  * Logging callback - allows the application to output log messages
-	  generated by librdkafka.
-  * partitioner callback - application provided message partitioner.
-	  The partitioner may be called in any thread at any time, it may be
-	  called multiple times for the same key.
-	  Partitioner function contraints:
-	  * MUST NOT call any rd_kafka_*() functions
-      * MUST NOT block or execute for prolonged periods of time.
-      * MUST return a value between 0 and partition_cnt-1, or the
-          special RD_KAFKA_PARTITION_UA value if partitioning
-              could not be performed.
+    An application typically does not have to perform any action when
+    an error is raised through the error callback, the client will
+    automatically try to recover from all errors, given that the
+    client and cluster is correctly configured.
+    In some specific cases a fatal error may occur which will render
+    the client more or less inoperable for further use:
+    if the error code in the error callback is set to
+    `RD_KAFKA_RESP_ERR__FATAL` the application should retrieve the
+    underlying fatal error and reason using the `rd_kafka_fatal_error()` call,
+    and then begin terminating the instance.
```

Comment:
In the bindings, I think this means we want a new error type `ErrorEvent` which is a subclass of `Error` and provides an additional `bool` property `IsFatal`. It doesn't seem like we want a more fine-grained property than this to indicate severity level.

Reply:
Yep, that's pretty much what the C++ API looks like: it combines log_cb, stats_cb, throttle_cb and error_cb into an Event object.
The Error Event has a fatal() a method.

Fatal errors are basically exceptions: ERR__FATAL indicates that the a fatal error has been raised, you then use rd_kafka_fatal_error() to extract the actual fatal error.
We should abstract in the most idiomatic way in the high-level clients, probably through exceptions raised by poll()/flush()/consume()/etc for .NET and Python.
For Go we'll keep using KafkaError but expose a fatal() method.